### PR TITLE
Add borderless select variant

### PR DESCRIPTION
**Pull request by FlyersPh9**
### CSS
- Add `data-iui-variant="borderless"` option.

### React
- Added `styleType: "default" | "borderless"` prop to native `Select` (`<select>`) and not the custom `Select` (`<div role="combobox">`)
  - This prop is excluded from `LabeledSelect` (https://github.com/iTwin/iTwinUI/pull/1886#discussion_r1516570250).
    - If passed to `LabeledSelect`, it is ignored.
- When `styleType="borderless"`:
  - No placeholder is allowed. If passed, it is ignored.
  - A defaultValue is required. If not passed, the first option is selected.

<details>
<summary>Outdated</summary>

> The big challenge was the types. I noticed a few type related potential issues and also some observations/demo cases. I added them to [App.tsx](https://gist.github.com/rohan-kadkol/b8eac88ce1e6050f0c137bf18eb7934d) GitHub Gist. Open to ideas on how to improve the types to fix the potential issues and/or thoughts if the issues pointed out are acceptable or must be fixed.

</details>

### CSS

- Added test for all of borderless, as well as hover and focus states.

### React

- Updated unit tests to fix type failures
- Added a unit test
- Added a type test
- Approved react-workshop images
- Tested in the vite-playground

## Docs

- Added a small section to the Select docs.
- Added changesets

## After PR TODOs

* [ ] Consider implementing the `defaultValue` prop in `CustomSelect` in case it is deemed as a bug (https://github.com/iTwin/iTwinUI/pull/1886#discussion_r1516849040).

### CODE REVIEW

File: apps/css-workshop/backstop/results/bitmaps_reference/iTwinUI_select_State_hover_light_0_demo-borderless-singular_0_.png
Comment (FlyersPh9) [inline apps/css-workshop/backstop/results/bitmaps_reference/iTwinUI_select_State_hover_light_0_demo-borderless-singular_0_.png:1]: Not sure why the hover tests are not showing the hover state. 🤔

File: packages/itwinui-css/src/select/select.scss [hunk 1]
```diff
@@ -74,6 +74,29 @@
...
+  --_iui-input-border-color: transparent;
+  transition: background-color var(--iui-duration-1) ease-out;
+
+  &:where(:hover, :active) {
+    --_iui-input-background-color: var(--iui-color-background-transparent-hover);
+    --_iui-input-border-color: transparent;
+    --_iui-input-with-icon-hover-border-color: transparent;
+  }
+
+  &:focus-visible {
```
Comment (FlyersPh9) [inline packages/itwinui-css/src/select/select.scss:88]: Did this to match with our pre-existing borderless buttons.

File: packages/itwinui-css/src/select/select.scss [hunk 2]
```diff
@@ -74,6 +74,29 @@
...
+    outline-offset: -1px;
+    outline-width: 1px;
+  }
+
+  &:is(.iui-disabled, :disabled, [aria-disabled='true']) {
```
Comment (FlyersPh9) [inline packages/itwinui-css/src/select/select.scss:93]: Did this to match with our pre-existing borderless buttons… however, I think we should discuss this as a team.  

Buttons do not have an empty / placeholder state, so a disabled button with no background color looks fine.
Selects DO have an empty / placeholder state, so the placeholder & disabled states are almost indistinguishable.  Thinking about keeping the default disabled state.

**Review by mayank99 — COMMENTED** [review]
I've merged #1879 into `main` and attempted to resolve merge conflicts in this PR. Hopefully I got it right, but if not I can revert it.

**Review by Ben-Pusey-Bentley — APPROVED** [review]
Looks good to me overall, I think that we should add an example of this to the Variants section of the Select Page for the documentation website.

File: packages/itwinui-react/src/core/Select/Select.tsx [hunk 3]
```diff
@@ -241,6 +243,12 @@ type SelectCommonProps = {
    * Modify size of select.
    */
   size?: 'small' | 'large';
+  /**
+   * Style of the select.
+   * Use 'borderless' to hide outline.
+   * @default 'default'
+   */
+  styleType?: 'default' | 'borderless';
```
Comment (mayank99) [inline packages/itwinui-react/src/core/Select/Select.tsx:251]: Should we make it more restrictive?
- We could allow borderless only for native select
- We could require passing a default selected value, and disallow placeholder

We can always make it less restrictive later if we learn about more use cases.

File: .changeset/lucky-dancers-smoke.md
Comment (mayank99) [inline .changeset/lucky-dancers-smoke.md:1]: Missing proper docs.

File: .changeset/lucky-dancers-smoke.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"@itwin/itwinui-react": minor
+---
+
+Added a `styleType` prop to `<Select>`. The possible values are `'default'` (default prop value) and `'borderless'`.
```
Comment (mayank99) [inline .changeset/lucky-dancers-smoke.md:5]: nit: This changeset could be reworded to center focus around "borderless". The "default" value is unimportant.

File: apps/css-workshop/pages/select.html [hunk 5]
```diff
@@ -503,6 +503,383 @@ <h2>Default</h2>
...
     <br />
 
+    <h2>Borderless</h2>
+    <section id="demo-borderless">
+
+      <label
+        class="iui-input-grid"
+        id="demo-borderless-singular"
+      >
+        <div class="iui-input-label iui-required">Select label</div>
```
Comment (mayank99) [inline apps/css-workshop/pages/select.html:513]: If LabeledSelect supports it, we should have a story for it too.

Also, we should think about the best way to label this and document it (can be done in the future if we're unsure). We are not even showing a label in https://github.com/iTwin/appui/pull/733

image

Comment (rohan-kadkol) [reply apps/css-workshop/pages/select.html:513]: > If LabeledSelect supports it, we should have a story for it too.

The reasoning was that although we support it, we don't recommend users to overuse it. So, we thought to not include a story for this in `LabeledSelect` but just in `Select`.

Comment (mayank99) [reply apps/css-workshop/pages/select.html:513]: In that case we should maybe exclude this prop from `LabeledSelect`, similar to other restrictions (mentioned in https://github.com/iTwin/iTwinUI/pull/1886#discussion_r1511207171).

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 3 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/Select/Select.tsx:251]: Some more thoughts on placeholder vs default selected value: Placeholder text can be confusing and is generally bad from an accessibility/usability perspective ([read more](https://www.smashingmagazine.com/2018/06/placeholder-attribute/)). In the specific use case that this PR is trying to address, there is no visible label either, which adds to the potential confusion. The default selected value is what provides the necessary context.

Comment (FlyersPh9) [reply packages/itwinui-react/src/core/Select/Select.tsx:251]: After our discussion about this, I am leaning towards requiring a value to be selected.  This removes the issue of https://github.com/iTwin/iTwinUI/pull/1886#discussion_r1507946815 .

File: apps/css-workshop/pages/select.html (hunk 5 above)
Comment (FlyersPh9) [reply apps/css-workshop/pages/select.html:513]: Let's exclude the prop from `LabeledSelect`.

File: .changeset/lucky-dancers-smoke.md (hunk 4 above)
Comment (FlyersPh9) [reply .changeset/lucky-dancers-smoke.md]: Updated.

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 3 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx]: Passing a `defaultValue` is now required in `CustomSelect` when `styleType="borderless"`. This is enforced in types. Additionally, if the user does not pass a `defaultValue`, the first option is automatically chosen.

Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx]: When `styleType="borderless"` in:

* `CustomSelect`: `defaultValue` is required (first option automatically selected if not passed) and `placeholder` is not allowed (ignored if passed). 
* `NativeSelect`: `defaultValue` is not required since anyways the first option was automatically selected when no placeholder or defaultValue was passed. So, I just disallowed `placeholder` (ignored if passed).

File: apps/css-workshop/pages/select.html (hunk 5 above)
Comment (rohan-kadkol) [reply apps/css-workshop/pages/select.html]: `styleType` is excluded from `LabeledSelect` in types and in code ([one of the commits](https://github.com/iTwin/iTwinUI/pull/1886/commits/d9895da725c582554c9e3e8183b536081ce807bf)).

File: packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.tsx [hunk 6]
```diff
@@ -150,9 +150,12 @@ export const LabeledSelect = React.forwardRef(
           disabled={disabled}
           className={className}
           style={style}
-          {...{ required: props.native ? required : undefined }}
-          {...rest}
           ref={forwardedRef}
+          {...({
+            required: props.native ? required : undefined,
+            ...rest,
+          } as SelectProps<T>)}
```
Comment (mayank99) [inline packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.tsx:157]: What is going on here?

File: packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.types-test.tsx [hunk 7]
```diff
@@ -20,7 +20,7 @@ import { LabeledSelect } from './LabeledSelect.js';
           { value: 2, label: 'Option 2' },
           { value: 3, label: 'Option 3' },
         ]}
-        onChange={(value) => {
+        onChange={(value: number) => {
```
Comment (mayank99) [inline packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.types-test.tsx:23]: Why is this necessary? It should be automatically inferred.

File: packages/itwinui-react/src/core/Select/Select.tsx [hunk 8]
```diff
@@ -267,13 +286,29 @@ const CustomSelect = React.forwardRef((props, forwardedRef) => {
...
+      // If borderless, empty value is not allowed. So, select the first option.
+      if (styleType === 'borderless') {
+        return isMultipleEnabled(valueProp, multiple)
+          ? [options[0]?.value]
+          : options[0]?.value;
+      }
+
+      return undefined;
+    })(),
+  );
```
Comment (mayank99) [inline packages/itwinui-react/src/core/Select/Select.tsx:311]: I believe we didn't previously support `defaultValue` in CustomSelect, so all this is not necessary for the purpose of this PR.

Also see [related thread](https://github.com/iTwin/iTwinUI/pull/1886#discussion_r1516798133)

File: packages/itwinui-react/src/core/Select/Select.tsx [hunk 9]
```diff
@@ -514,21 +548,33 @@ export type CustomSelectProps<T> = SelectCommonProps & {
...
+       * Use 'borderless' to hide outline.
+       * @default 'default'
+       */
+      styleType?: 'default';
+      /**
+       * Placeholder when no item is selected.
+       *
+       * When `styleType=borderless`, `placeholder` is not allowed. Additionally, a `defaultValue` must be provided.
+       */
+      placeholder?: React.ReactNode;
```
Comment (mayank99) [inline packages/itwinui-react/src/core/Select/Select.tsx:577]: nit: I think the jsdocs are now repeated and could be removed in certain places?

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 3 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx]: That will simplify things and might also not cause the type issues (in the GitHub gist) that I mentioned in the PR description.

I hadn't done this so far because Jon didn't say so in his reply. Also, the CSS styles also handled borderless for the `CustomSelect` case. So, I thought we all agreed to support it for `CustomSelect` too.

@FlyersPh9 Can/Should we support `borderless` only on `NativeSelect`?

Comment (mayank99) [reply packages/itwinui-react/src/core/Select/Select.tsx]: Btw you might want to move the PR to draft next time until it is ready. I finished my review before seeing your updated PR description 😅

image

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 9 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx:577]: Yes, maybe. Actually, I was hoping we address it later along with https://github.com/iTwin/iTwinUI/pull/1862#discussion_r1492672122. That's why I didn't test the JSDocs too much and just concentrated on the logic.

File: packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.tsx (hunk 6 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.tsx:157]: When doing `Omit<SelectProps<T>, "styleType">` instead of `SelectProps<T>`, I was getting strange type behaviors and errors. This type assertion fixes (ignores) the error. But open to proper ways to fix it:

<details>
<summary>Error screenshots</summary>

image
image

</details>

File: packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.types-test.tsx (hunk 7 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.types-test.tsx:23]: Similar to https://github.com/iTwin/iTwinUI/pull/1886#discussion_r1516824519, doing `Omit<SelectProps<T>, "styleType">` instead of `SelectProps<T>` gave strange type behaviors and errors. That comment was the resulting error. This comment is the resulting strange behavior.

I made a mention of this case in the GitHub gist that I had included in the PR description. I tried quite a few approaches to fix the "potential issues" I mentioned there, but I couldn't fix the last remaining ones, which included this issue of `onChange`'s `value` argument not getting the inferred type. Open to different approaches.

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 8 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx:311]: Yes, we didn't support it previously for `CustomSelect`. I thought that was a bug and so I fixed it here.

Are you suggesting that this fix can come in a separare PR? Or are you suggesting reverting this altogether?

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 3 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx]: Ohh, sounds good, will keep that in mind. Since I didn't update the PR description, I thought it was assumed that I'm working on the React portion and it's not complete and so not ready for review. But I'll mark the PR back to a draft next time to avoid confusion.

To confirm, the PR is now ready to review 🙂.

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 8 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/Select/Select.tsx:311]: It should be reverted, at least for now. This PR's main purpose is: "add `styleType="borderless"` support to `<Select native>`".

_If_ we want to support `defaultValue`, it should happen in a separate PR.

Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx]: Sounds good, reverted (https://github.com/iTwin/iTwinUI/pull/1886/commits/6eabe377fd4b159eb66a99a84d8a78d1c66e6322). Added this comment to the after-PR TODOs.

File: packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.types-test.tsx (hunk 7 above)
Comment (mayank99) [reply packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.types-test.tsx:23]: Looks like `styleType?: never` works as a workaround.

Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.types-test.tsx]: Yes, after our recent commits (https://github.com/iTwin/iTwinUI/pull/1886/files/ae07a6fb42b4d464e50d42a3cd0347dc31c43da0..aca5cf10a4c6b7504e9eae54b630913074789f41), that highlighted change was no longer needed, and thus reverted.

File: packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.tsx (hunk 6 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/LabeledSelect/LabeledSelect.tsx]: After our recent commits (https://github.com/iTwin/iTwinUI/pull/1886/files/ae07a6fb42b4d464e50d42a3cd0347dc31c43da0..aca5cf10a4c6b7504e9eae54b630913074789f41), that type assertion was removed.

**Comment by rohan-kadkol** [comment]
> I think that we should add an example of this to the Variants section of the Select Page for the documentation website.

Sounds good, added (https://github.com/iTwin/iTwinUI/pull/1886/commits/2b5968737070bba7bdbee74a31ac355b55e10d38)

File: apps/react-workshop/cypress-visual-screenshots/baseline/Select.test.ts-Borderless (Open).png
Comment (mayank99) [inline apps/react-workshop/cypress-visual-screenshots/baseline/Select.test.ts-Borderless (Open).png:1]: You'll probably have to change `Select.test.ts` to early exit for this particular story because it is not possible to test the "open" state of native Select.

**Review by mayank99 — APPROVED** [review]
LGTM!

Only the test images need to be updated before merging.

File: packages/itwinui-react/src/core/Select/Select.tsx (hunk 3 above)
Comment (rohan-kadkol) [reply packages/itwinui-react/src/core/Select/Select.tsx]: After a short internal discussion, we decided to only support borderless for the native select (`<select>`) and not the custom select (`<div role="combobox">`).

File: apps/react-workshop/cypress-visual-screenshots/baseline/Select.test.ts-Borderless (Open).png
Comment (rohan-kadkol) [reply apps/react-workshop/cypress-visual-screenshots/baseline/Select.test.ts-Borderless (Open).png:1]: Thanks, done (https://github.com/iTwin/iTwinUI/pull/1886/commits/0fa7c18e39270c927a7310d0e97a8ae5ccaf53cc).

**Comment by Juliakas** [comment]
@FlyersPh9 @mayank99 Sorry for commenting on this 1 year old PR, but just stumbled upon it from comment on source code: 
image

Why not allow bordlerless to be set on CustomSelect? When I bypass typescript checks I get perfectly working component like this:
image

Also considering that we have other borderless buttons that open dropdown menus it fits in well.

**Comment by mayank99** [comment]
@Juliakas This restriction was placed because it was more difficult to cover all the edge cases when implementing a borderless custom select; and the original use-case for this feature was covered by the native select. I believe your use-case should also covered by native select. Other than that, it's worth noting that in general, borderless controls should be carefully and sparingly used.

If you have any further questions or comments, please start a new [discussion](https://github.com/iTwin/iTwinUI/discussions).
